### ISSUE TITLE
RFE: create a shared key macro capability

### ISSUE THREAD

**Issue by i30817** [libretro/RetroArch#7775]
## Description

If the creation and saving of macros per game (it really only makes sense per game so you can save automatically there) was possible in retroarch, a bit like zsnes, some games would much easier to do cool things on.

### Expected behavior

A menu like zsnes had where you could create a macro without requiring to actually input the tricky move yourself in real time. 
If i recall correctly, zsnes presented a menu with all possible keys and allowed you to click them into a sequence, or click on a 'wait' marker which would get presented on a label. 

I think this can be made better for retroarch (which often does not support mouses in some configurations) by instead:

1. presenting a image of the controller
2. wait for user input
3. advance time (frame or more i guess) or exit on two 'unusual' key combinations (if you want fully controller controlled key macro configuration). They'd have no effect outside of this key recorder widget
4. show the buttons that are recorded on the macro on the controller image and repeat as it goes on, maybe with a clear marker for start and end.

On the input menu you should be able to assign keys to macros, which do nothing until connected to a macro on the new macro menu. These keys would be 'special' and if you keep pressing them they do nothing until the frame the macro 'ends'. Similarly, if another game or macro key is pressed, the current macro is interrupted and the other takes over.

On the macro menu you'd be able to set a new macro to one of the key macros.

Some keys are probably illegal to press at the same time (like different axis for example) but those are details.

You probably would be able to reuse the 'key sending' capability of multiplayer for injecting a macro on the controller.

I don't know if this is the best design, and it's probably a lot of work, but it seems a cool idea, so here it is for posterity or glory.

### Actual behavior

Nothing, feature is not available, though it probably could exist on all cores, in emulators that never had the capability.

**Comment by i30817** [comment]
If some one read this already i edited the macro capture section because i had a better idea than key combinations to control it (keeping buttons pressed to control, quick tap to select).

**Comment by i30817** [comment]
if you make the copy paste macro somewhat readable people will also be able to make minimal macros manually (except for analog movement again i guess).

In fact that could be a 'minimal' way to implement the feature without any other capture UI / analog movement complications.

**Comment by i30817** [comment]
Ok, i got what's probably a better, less ambitious idea. I wrote it up there in 'Alternative: no gui macro creation', but if it's tl;dr, basically instead of trying to make users able to record their own macros, create a protocol describing a macro (and splines operating on macro intervals for the analog movements) and a paste from clipboard button.

That's it, leave the nerds to create the complicated spline wii-mote movement macros, or reasonably diligent people that can chain together a list for the normal macros and leave the lusers to get them from copy pasting from the internet.

This would be a good first iteration because getting this right could be seen as a step for the more complicated input capture the rest of the post attempts to describe.

**Comment by fr500** [comment]
The input remapper would certainly allow this to happen.

But it is far from trivial to implement.
Combos may be achievable within a reasonable timeframe (sans GUI) but this is in a whole different level of complexity.

**Comment by i30817** [comment]
I agree, but i'm rooting for the limited version i put in my last post. You don't *really* need to make a GUI for the functionality to be useful, just for opening the door to 'anyone' creating one without a text editor.

If you create a parser that takes something like:
"Ryu hadoken" [down;right-down;right; x]"

and creates a macro entry, that would be a very good first step. Though hopefully most fightgames don't mirror combos, that i have no clue how to deal with without duplication.
